### PR TITLE
feat: add ability to use custom tunnel in dev cmd

### LINKED ISSUE #590: [TRI-1390] feat: Add ability to use custom tunnel in `dev` command
### Is your feature request related to a problem? Please describe.

If I already have an ngrok free tunnel running pointing to my local dev server that I want to use with Trigger.dev, the CLI `dev` command fails with the following message:

```sh
✖ Ngrok failed to create a tunnel for port 3000 because ngrok is already running
```

### Describe the solution you'd like to see

I'd like to be able to provide a flag to the `dev` CLI command to pass in my already running tunnel URL:

```sh
npx @trigger.dev/cli dev --tunnel https://abcdefg.ngrok.io
```

### Describe alternate solutions

The only alternative currently is to subscribe to an ngrok paid plan

### Additional information

_No response_

<sub>[TRI-1390](https://linear.app/triggerdotdev/issue/TRI-1390/feat-add-ability-to-use-custom-tunnel-in-dev-command)</sub>

### PR DESCRIPTION
**Pull request by hmacr**
Closes #590 

## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/triggerdotdev/trigger.dev/blob/main/CONTRIBUTING.md)
- [x] The PR title follows the convention.
- [x] I ran and tested the code works

---

## Testing

- Verified that the tunnel-url is used to connect to the server.
- Verified that the endpoint verification fails if the tunnel-url is invalid, and the `dev` command proceeds to check the next possible endpoint.
- Verified that the `dev` command creates a new tunnel if tunnel-url is not provided.

---

## Changelog

- Added `--tunnel` option to the `dev` command that accepts either a `http` or `https` URL.
- Verify that the provided tunnel-url points to the hosted server.
- Use the provided tunnel-url instead of creating a new ngrok tunnel.

---

💯

### CODE REVIEW

File: packages/cli/src/commands/dev.ts [hunk 1]
```diff
@@ -294,6 +303,30 @@ async function verifyEndpoint(
       }
 
       spinner.succeed(`[trigger.dev] Found your trigger endpoint: ${localEndpointHandlerUrl}`);
+
+      if (resolvedOptions.tunnel) {
```
Comment (ericallam) [inline packages/cli/src/commands/dev.ts]: Remove this in favor of using the existing `verifyEndpoint` function which does this exact thing,

Comment (hmacr) [reply packages/cli/src/commands/dev.ts]: This happens inside the `verifyEndpoint` function. This is done as an extra check that verifies the provided tunnel-url points to the resolved URL. I'm assuming that we still want the `dev` command to resolve possible URLs when a custom tunnel-url is provided.

Flow (as of now): Based on framework, get possible URLs -> verify the URL hosts a valid `trigger` endpoint -> If tunnel-url is provided, verify the tunnel-url points to the resolved URL for the framework.

But, I see that skipping the framework-based URL resolution and using the `tunnel-url` directly (when available), makes complete sense. Let me modify the PR.

**Comment by ericallam** [comment]
@matt-aitken is wrapping up some changes to the dev CLI command and will pick up this PR after that (hopefully later today)

**Comment by hmacr** [comment]
@matt-aitken I've rebased the PR with the main branch, please review.

**Review by matt-aitken — APPROVED** [review]
This is brilliant, it works great. Thanks @hmacr

**Pull request merged**